src/apply.c

```diff
@@ -376,3 +417,364 @@ int git_apply__patch(
 
 	return error;
 }
+
+static int apply_one(
+	git_repository *repo,
+	git_reader *preimage_reader,
+	git_index *preimage,
+	git_index *postimage,
+	git_diff *diff,
+	size_t i,
+	const git_apply_options *opts)
+{
+	git_patch *patch = NULL;
+	git_buf pre_contents = GIT_BUF_INIT, post_contents = GIT_BUF_INIT;
+	const git_diff_delta *delta;
+	char *filename = NULL;
+	unsigned int mode;
+	git_oid pre_id, post_id;
+	git_index_entry pre_entry, post_entry;
+	int error;
+
+	if ((error = git_patch_from_diff(&patch, diff, i)) < 0)
+		goto done;
+
+	delta = git_diff_get_delta(diff, i);
+
+	if (opts->delta_cb) {
+		error = opts->delta_cb(delta, opts->payload);
+
+		if (error) {
+			if (error > 0)
+				error = 0;
+
+			goto done;
+		}
+	}
+
+	if (delta->status != GIT_DELTA_ADDED) {
+		error = git_reader_read(&pre_contents, &pre_id,
+			preimage_reader, delta->old_file.path);
+
+		/* ENOTFOUND means the preimage was not found; apply failed. */
+		if (error == GIT_ENOTFOUND)
+			error = GIT_EAPPLYFAIL;
```

Comment:
I do overwrite `GIT_ENOTFOUND`, but I don't overwrite the error message itself, which would have been set by the reader.  This means that it's actually more explicit, since the existing error message distinguishes between the index and the working directory.  eg, if it's not found in the index, it will say [`index does not contain 'foo'`](https://github.com/libgit2/libgit2/blob/master/src/index.c#L867).